### PR TITLE
Add ip filter design doc

### PR DESCRIPTION
**Pull request by ecordell**
I took a stab at a proposal to address https://github.com/projectcontour/contour/issues/3693

I haven't looked deeply at the envoy filters available for this, but I wrote down what looked like would work based on the docs.

### CODE REVIEW

**Review by sunjayBhatia — COMMENTED** [review]
it would be great to add some implementation notes about how the Envoy filters might work, once we agree on a general shape of the API

i.e. mention the RBAC filter: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rbac_filter and how configuring Prinicipals might work: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#envoy-v3-api-msg-config-rbac-v3-principal

some examples of configuration and requests (including source ip and/or x-forwarded-for header) that would be allowed/denied would be great to add

the Envoy filter allows full per-route configuration so the per-route configuration makes a lot of sense to me

File: design/ip-filtering-design.md
Comment (sunjayBhatia) [inline design/ip-filtering-design.md:20]: we'll likely need to write up a design for a CR to be used as an extension Filter or Policy for Gateway API

File: design/ip-filtering-design.md
Comment (sunjayBhatia) [inline design/ip-filtering-design.md:71]: writing down how exactly these concepts fold into the Envoy RBAC configuration fields would be good to make sure to head off any confusion early

**Comment by ecordell** [comment]
Thanks for the review @sunjayBhatia - I'll work on ironing out the envoy details and update here.

File: design/ip-filtering-design.md
Comment (ecordell) [reply design/ip-filtering-design.md:20]: Are you saying that you think we should bring the Gateway API into scope? Or just that that's the likely direction if/when this is approached in for Gateway API?

Comment (sunjayBhatia) [reply design/ip-filtering-design.md:20]: a bit of both actually, probably good to get a baseline of what we want to support done in HTTPProxy as well as acknowledging that unless the upstream Gateway API project supports it natively we will have to write our own CR to handle that feature

**Review by tsaarni — COMMENTED** [review]
Great design doc @ecordell, Thank You for working on this! I've added some comments/questions.

File: design/ip-filtering-design.md
Comment (ecordell) [reply design/ip-filtering-design.md:71]: Yep, I took that from Ambassador (it was linked in one of the original GH issues on ip filtering). I don't have a strong opinion on the fields names, if `direct_remote` and `remote` sound better.

**Comment by ecordell** [comment]
@sunjayBhatia @tsaarni I addressed most of the feedback and made a couple of updates to the proposed API.

I've implemented this https://github.com/projectcontour/contour/pull/5008 with the understanding that we may need to adjust it if we change anything here first. Implementing it was the easiest way to suss out exactly how envoy's rbac filter actually works in practice.

File: design/ip-filtering-design.md
Comment (ecordell) [reply design/ip-filtering-design.md:39]: I can test this - I'm not 100% sure what happens, but I would defer to whatever envoy does here.

**Comment by ecordell** [comment]
Since I think the API design is what probably requires the most discussion, here are some alternatives to think about:

1. What's in the proposal right now:

```yaml
      ipAllowPolicy:
      - cidr: 127.0.0.1/32
        source: Peer
      #or 
      ipDenyPolicy:
      - cidr: 99.99.0.0/16
        source: Remote
```
- separate top-level fields for allow vs. deny
- separate field for cidr range and ip source
- uses `Remote` / `Peer` terminology instead of `Remote` and `DirectRemote`

2. An option with only one top-level field: 

```yaml
      ipFilterPolicy:
      - allow: 127.0.0.1/32
        source: Peer
      - allow: 99.99.0.0/16
        source: Peer
      #or 
      ipFilterPolicy:
      - deny: 99.99.0.0/16
        source: Remote
```
- single top-level field for filters, each filter specifies allow vs. deny
- separate field for cidr range and ip source
- uses `Remote` / `Peer` terminology instead of `Remote` and `DirectRemote`

3. Using envoy terminology for remotes:

```yaml
      ipFilterPolicy:
      - allow: 127.0.0.1/32
        source: DirectRemote
      - allow: 99.99.0.0/16
        source: Remote
      #or 
      ipFilterPolicy:
      - deny: 99.99.0.0/16
        source: Remote
```
- single top-level field for filters, each filter specifies allow vs. deny
- separate field for cidr range and ip source
- uses `Remote` and `DirectRemote`

4. If the top-level field indicates allow vs. deny, we can get by with only one field for the actual filters:

```yaml
      ipAllowPolicy:
      - directRemote: 127.0.0.1/32
      - remote: 99.99.0.0/16
      #or 
      ipDenyPolicy:
      - remote: 99.99.0.0/16
```

- separate top-level fields for allow vs. deny
- single field for ip source and range

Another api question is weather bare ips should be interpreted as `/32` cidrs. Right now I have everything written as an explicit cidr range.

File: design/ip-filtering-design.md
Comment (sunjayBhatia) [reply design/ip-filtering-design.md:39]: based on how we program match conditions the more specific prefix will get the routing priority, so since we're using per-route filters the IP policy should be applied after the route is chosen

**Comment by sunjayBhatia** [comment]
> Another api question is weather bare ips should be interpreted as /32 cidrs. Right now I have everything written as an explicit cidr range.

I'm in favor of being more explicit, I don't think it's too bad to ask people to add `/32` or `/128` if they want to allow/deny a particular IP

**Comment by sunjayBhatia** [comment]
Also I'm somewhat in favor of the options that have only one top-level field, just so there are less keys to have to configure and people don't get the impression just looking at the shape of the API that they could combine allow and deny policies on a route

**Comment by ecordell** [comment]
> I'm in favor of being more explicit, I don't think it's too bad to ask people to add /32 or /128 if they want to allow/deny a particular IP

I tend to agree, but for what it's worth the nginx whitelist feature supports single-ips in the filter.

**Comment by izturn** [comment]
@ecordell  any progress?

**Comment by tsaarni** [comment]
> > I'm in favor of being more explicit, I don't think it's too bad to ask people to add /32 or /128 if they want to allow/deny a particular IP
> 
> I tend to agree, but for what it's worth the nginx whitelist feature supports single-ips in the filter.

`iptables -s address[/mask]` also allows using single IP address, from [man page](https://man7.org/linux/man-pages/man8/iptables.8.html):  "_Address can be either a network name, a hostname, a network IP address (with /mask), or a plain IP address._". It seems to be pretty common shorthand with some routers as well (like [here](https://www.juniper.net/documentation/us/en/software/junos/security-policies/topics/topic-map/security-policy-configuration.html#id-dynamic-address-groups-in-security-policies)).

**Comment by ecordell** [comment]
@tsaarni I agree, allowing bare IPs probably makes the most sense. 

@izturn I believe I've addressed all of the feedback in this PR; I'm just waiting for further review and approval. Is there something specific you'd like to see addressed?

**Comment by izturn** [comment]
@ecordell, thx for your work, I can't wait to use it, btw I saw there are two branches: `ipfilter`, `annotation-ipfilter` in your fork, which one is your prefer?

**Comment by ecordell** [comment]
Thanks @izturn!

`ipfilter` is the branch we should look at - `annotation-ipfilter` is a fork that uses annotations to drive the ipfiltering so that we can use the ipfiltering features before this proposal is approved and merged.

**Comment by izturn** [comment]
got it

File: design/ip-filtering-design.md
Comment (wilsonwu) [inline design/ip-filtering-design.md:38]: Could we support allow and deny IP list for virtualhost level?
```suggestion
    fqdn: foo-basic.bar.com
    ipAllowPolicy:
      # traffic is allowed if it came from localhost (i.e. co-located reverse proxy)
      - cidr: 127.0.0.1/32
        source: Peer
      # and the request originated from an IP in this range 
      - cidr: 99.99.0.0/16
        source: Remote
```

Comment (izturn) [reply design/ip-filtering-design.md:38]: @ecordell

**Comment by ecordell** [comment]
@sunjayBhatia I'll work on updating for virtualhost filtering, sorry for the silence

**Comment by sunjayBhatia** [comment]
> Late to the party on this one. Is supporting the `ingress.kubernetes.io/whitelist-source-range` annotation on ingress objects in the scope of this design? If we are adding IP filtering then it would be great to be a drop in replacement for other ingress replacements. That annotation is the only reason I keep an nginx controller running in our clusters.

not at the moment, no, but if you're willing to contribute an addendum to this document or an additional feature once this is merged/agreed upon we will of course consider it

**Comment by ecordell** [comment]
@sunjayBhatia doc is updated for virtualhost-level ip filtering

It's implemented in https://github.com/projectcontour/contour/pull/5008 as well

**Review by wilsonwu — APPROVED** [review]
lgtm

File: design/ip-filtering-design.md
Comment (izturn) [inline design/ip-filtering-design.md:17]: From the change log:
>Filters can indicate whether the direct IP should be used or whether a reported IP from `PROXY` or `X-Forwarded-For` should be used instead.

so what's about `PROXY`?

File: design/ip-filtering-design.md
Comment (izturn) [inline design/ip-filtering-design.md:88]: `Peer`/`Remote` OR `direct_remote_ip`/`remote_ip` ?

File: design/ip-filtering-design.md
Comment (izturn) [reply design/ip-filtering-design.md]: > Is the terminology `peer` and `remote` originating from Ambassador? Envoy seems to use [`direct_remote_ip`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#envoy-v3-api-field-config-rbac-v3-principal-direct-remote-ip) and [`remote_ip`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#envoy-v3-api-field-config-rbac-v3-principal-remote-ip). Which one is better to align?

I like the latter

**Review by izturn — APPROVED** [review]
/lgtm

File: design/ip-filtering-design.md
Comment (ecordell) [reply design/ip-filtering-design.md:17]: > maybe we need to unify them all

Not sure I'm following your suggestion, would you mind rephrasing it?

Comment (izturn) [reply design/ip-filtering-design.md:17]: sorry for my pool english, i means the descriptions in these two places need to be consistent

File: design/ip-filtering-design.md
Comment (sunjayBhatia) [inline design/ip-filtering-design.md:3]: can switch this to approved

**Review by sunjayBhatia — APPROVED** [review]
Taking this and the implementation in #5008 together I think this is good to approve, the open questions I had that weren't explicitly covered here were around how exactly the overrides would work (virtualhost config overridden per-route) and how includes would work, but looks like that is already tested in the implementation so we can discuss there

I'll leave this open for a little longer in case anyone else wants to chime in

**Review by sunjayBhatia — COMMENTED** [review]
one more thought I just had would be to nail down what the precedence of the RBAC filter is when combined with other filters

i.e. should IP filtering happen before rate limiting, ExtAuth etc.

since the filter ends up getting applied as a per-route filter this might mean it is evaluated later than people would expect

File: design/ip-filtering-design.md
Comment (skriss) [inline design/ip-filtering-design.md:27]:
```suggestion
The `virtualhost` and `route` sections of an `HTTPProxy` will each have two new fields added: `ipAllowPolicy` and `ipDenyPolicy`.
```

**Review by skriss — COMMENTED** [review]
One small clarification for posterity

**Pull request merged**
